superdesk/publish/formatters/ninjs_ftp_formatter.py

```diff
@@ -47,48 +42,212 @@ def _transform_to_ninjs(self, article, subscriber, recursive=True):
         :param recursive:
         :return:
         """
-
+        # Get the path that the renditions will be pushed to
+        self.path = subscriber.get("destinations")[0].get("config").get("associated_path")
         include_original = subscriber.get("destinations")[0].get("config").get("include_original", False)
         if include_original:
             self.internal_renditions = ["original"]
 
-        ninjs = super()._transform_to_ninjs(article, subscriber, recursive)
+        formatted_article = deepcopy(article)
 
-        # Get the path that the renditions will be pushed to
-        path = subscriber.get("destinations")[0].get("config").get("associated_path")
+        if article.get("type") == "text" and recursive:
+            self.apply_product_filtering_to_associations(formatted_article, subscriber)
+
+        ninjs = super()._transform_to_ninjs(formatted_article, subscriber, recursive)
+
+        renditions = ninjs.get("renditions")
+        if renditions:
+            for name, rendition in renditions.items():
+                rendition["href"] = (
+                    self.path.lstrip("/")
+                    + ("/" if not self.path.endswith("/") and self.path else "")
+                    + get_rendition_file_name(rendition)
+                )
+
+        return ninjs
+
+    def apply_product_filtering_to_associations(self, article, subscriber):
+        """
+        Remove the embedded items from the article that the subscriber has no matching product for.
+        :param article:
+        :param subscriber:
+        :return:
+        """
+        kill_keys = []
+        permitted_products = set(subscriber["products"])
+
+        for key, item in article.get("associations", {}).items():
+            if key.startswith("editor_"):
+                result = get_resource_service("product_tests").test_products(item, lookup=None)
+                matching_products = set(p["product_id"] for p in result if p.get("matched", False))
+                if not matching_products.intersection(permitted_products):
+                    kill_keys.append(key)
+
+        self.remove_embeds(article, kill_keys)
+
+    def remove_embeds(self, article, kill_keys):
```

Comment:
A lot of similarities here between this and the parent `remove_all_embeds` function.

Would it be possible if the `remove_all_embeds` (once moved to `editor_utils`) was a more generic version that allows to provide a custom filter function/side-affects instead?

Reply:
Not that keen! the remove_all_embeds is a sweet simple thing. What's going on the the formatter is more complex, I'd prefer not to mix them.